### ISSUE TITLE
support for multiple keys in sampling functions

### ISSUE THREAD

**Issue by deanmarchiori** [njtierney/brolgar#89]
When using `sample_n_keys()` on a tsibble with one 'key' the function works as expected. For a tsibble with 2 keys, the function returns an error. Expected behaviour would probably be sampling on the subset of data uniquely identied by both (or more) keys. 

I haven't digged into the source so not sure if it's feasible to extend, but happy to help out on a fix if this is something you think would add value?

``` r
library(tsibble)
library(brolgar)
library(tsibbledata)

data("aus_retail")

# has multiple keys
aus_retail
#> # A tsibble: 64,532 x 5 [1M]
#> # Key:       State, Industry [152]
#>    State             Industry                `Series ID`     Month Turnover
#>    <chr>             <chr>                   <chr>           <mth>    <dbl>
#>  1 Australian Capit… Cafes, restaurants and… A3349849A    1982 Apr      4.4
#>  2 Australian Capit… Cafes, restaurants and… A3349849A    1982 May      3.4
#>  3 Australian Capit… Cafes, restaurants and… A3349849A    1982 Jun      3.6
#>  4 Australian Capit… Cafes, restaurants and… A3349849A    1982 Jul      4  
#>  5 Australian Capit… Cafes, restaurants and… A3349849A    1982 Aug      3.6
#>  6 Australian Capit… Cafes, restaurants and… A3349849A    1982 Sep      4.2
#>  7 Australian Capit… Cafes, restaurants and… A3349849A    1982 Oct      4.8
#>  8 Australian Capit… Cafes, restaurants and… A3349849A    1982 Nov      5.4
#>  9 Australian Capit… Cafes, restaurants and… A3349849A    1982 Dec      6.9
#> 10 Australian Capit… Cafes, restaurants and… A3349849A    1983 Jan      3.8
#> # … with 64,522 more rows

# fails
aus_retail %>% 
  sample_n_keys(size =10)
#> Error in .subset2(x, i): subscript out of bounds


# subset of data and enforce jsut one key
onekey <- aus_retail %>% 
  dplyr::filter(State == "New South Wales") %>% 
  as_tsibble(key = Industry, index = Month) 
  
onekey
#> # A tsibble: 8,820 x 5 [1M]
#> # Key:       Industry [20]
#>    State        Industry                      `Series ID`    Month Turnover
#>    <chr>        <chr>                         <chr>          <mth>    <dbl>
#>  1 New South W… Cafes, restaurants and cater… A3349709X   1982 Apr     61.8
#>  2 New South W… Cafes, restaurants and cater… A3349709X   1982 May     60.8
#>  3 New South W… Cafes, restaurants and cater… A3349709X   1982 Jun     58.7
#>  4 New South W… Cafes, restaurants and cater… A3349709X   1982 Jul     60.3
#>  5 New South W… Cafes, restaurants and cater… A3349709X   1982 Aug     56.1
#>  6 New South W… Cafes, restaurants and cater… A3349709X   1982 Sep     58.1
#>  7 New South W… Cafes, restaurants and cater… A3349709X   1982 Oct     53.9
#>  8 New South W… Cafes, restaurants and cater… A3349709X   1982 Nov     61.2
#>  9 New South W… Cafes, restaurants and cater… A3349709X   1982 Dec     75.7
#> 10 New South W… Cafes, restaurants and cater… A3349709X   1983 Jan     54.2
#> # … with 8,810 more rows

# works fine
onekey %>% 
  sample_n_keys(size = 10)
#> # A tsibble: 4,410 x 5 [1M]
#> # Key:       Industry [10]
#>    State           Industry           `Series ID`    Month Turnover
#>    <chr>           <chr>              <chr>          <mth>    <dbl>
#>  1 New South Wales Clothing retailing A3349399C   1982 Apr     94  
#>  2 New South Wales Clothing retailing A3349399C   1982 May    106. 
#>  3 New South Wales Clothing retailing A3349399C   1982 Jun     95.1
#>  4 New South Wales Clothing retailing A3349399C   1982 Jul     95.3
#>  5 New South Wales Clothing retailing A3349399C   1982 Aug     82.8
#>  6 New South Wales Clothing retailing A3349399C   1982 Sep     89.4
#>  7 New South Wales Clothing retailing A3349399C   1982 Oct     83.3
#>  8 New South Wales Clothing retailing A3349399C   1982 Nov     99.3
#>  9 New South Wales Clothing retailing A3349399C   1982 Dec    143. 
#> 10 New South Wales Clothing retailing A3349399C   1983 Jan     78.5
#> # … with 4,400 more rows
```

<sup>Created on 2020-04-03 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by deanmarchiori** [comment]
connected to #85

**Comment by njtierney** [comment]
I'm hoping to get to work on this soon, but I would definitely welcome any assistance here, 😄

**Comment by deanmarchiori** [comment]
Taking a closer look here, it looks like PR #72 will address this. Were you thinking of merging that PR?

I have expanded on my example below with the new function. It looks like it's the use of `tsibble::key_rows()` rather than `tsibble::key_vars()`.


``` r
library(tsibble)
library(brolgar)
library(tsibbledata)

data("aus_retail")

# has multiple keys
aus_retail
#> # A tsibble: 64,532 x 5 [1M]
#> # Key:       State, Industry [152]
#>    State             Industry                `Series ID`     Month Turnover
#>    <chr>             <chr>                   <chr>           <mth>    <dbl>
#>  1 Australian Capit… Cafes, restaurants and… A3349849A    1982 Apr      4.4
#>  2 Australian Capit… Cafes, restaurants and… A3349849A    1982 May      3.4
#>  3 Australian Capit… Cafes, restaurants and… A3349849A    1982 Jun      3.6
#>  4 Australian Capit… Cafes, restaurants and… A3349849A    1982 Jul      4  
#>  5 Australian Capit… Cafes, restaurants and… A3349849A    1982 Aug      3.6
#>  6 Australian Capit… Cafes, restaurants and… A3349849A    1982 Sep      4.2
#>  7 Australian Capit… Cafes, restaurants and… A3349849A    1982 Oct      4.8
#>  8 Australian Capit… Cafes, restaurants and… A3349849A    1982 Nov      5.4
#>  9 Australian Capit… Cafes, restaurants and… A3349849A    1982 Dec      6.9
#> 10 Australian Capit… Cafes, restaurants and… A3349849A    1983 Jan      3.8
#> # … with 64,522 more rows

# will fail
aus_retail %>% 
  sample_n_keys(size = 10)
#> Error in .subset2(x, i): subscript out of bounds

# logic in PR72
sample_n_keys_new <- function(.data, size){
  key_indices <- tsibble::key_rows(.data)
  sample_unique_keys <- sample(key_indices, size)
  dplyr::slice(.data, vctrs::vec_c(!!!sample_unique_keys))
}

# success!
aus_retail %>% 
  sample_n_keys_new(size = 10)
#> # A tsibble: 4,266 x 5 [1M]
#> # Key:       State, Industry [10]
#>    State    Industry           `Series ID`    Month Turnover
#>    <chr>    <chr>              <chr>          <mth>    <dbl>
#>  1 Tasmania Clothing retailing A3349371A   1982 Apr      6.7
#>  2 Tasmania Clothing retailing A3349371A   1982 May      7.4
#>  3 Tasmania Clothing retailing A3349371A   1982 Jun      6.7
#>  4 Tasmania Clothing retailing A3349371A   1982 Jul      7.1
#>  5 Tasmania Clothing retailing A3349371A   1982 Aug      5.8
#>  6 Tasmania Clothing retailing A3349371A   1982 Sep      5.8
#>  7 Tasmania Clothing retailing A3349371A   1982 Oct      5.3
#>  8 Tasmania Clothing retailing A3349371A   1982 Nov      7.1
#>  9 Tasmania Clothing retailing A3349371A   1982 Dec     11.1
#> 10 Tasmania Clothing retailing A3349371A   1983 Jan      5.3
#> # … with 4,256 more rows
```

<sup>Created on 2020-08-18 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by njtierney** [comment]
Oh right! Nice! That's excellent. Thank you so much for taking a close look at this, really appreciate it, Dean.

I overlooked #72 because I got thinking about S3 generics and that side of things. I'll merge that in and add some tests.

**Comment by njtierney** [comment]
Thanks for this, @deanmarchiori , this is now resolved from #97. Some interesting bugs along the way - bit of a journey! Thanks for your help.

    library(tsibble)
    library(brolgar)
    library(tsibbledata)

    data("aus_retail")

    # has multiple keys
    aus_retail
    #> # A tsibble: 64,532 x 5 [1M]
    #> # Key:       State, Industry [152]
    #>    State               Industry                    `Series ID`    Month Turnover
    #>    <chr>               <chr>                       <chr>          <mth>    <dbl>
    #>  1 Australian Capital… Cafes, restaurants and cat… A3349849A   1982 Apr      4.4
    #>  2 Australian Capital… Cafes, restaurants and cat… A3349849A   1982 May      3.4
    #>  3 Australian Capital… Cafes, restaurants and cat… A3349849A   1982 Jun      3.6
    #>  4 Australian Capital… Cafes, restaurants and cat… A3349849A   1982 Jul      4  
    #>  5 Australian Capital… Cafes, restaurants and cat… A3349849A   1982 Aug      3.6
    #>  6 Australian Capital… Cafes, restaurants and cat… A3349849A   1982 Sep      4.2
    #>  7 Australian Capital… Cafes, restaurants and cat… A3349849A   1982 Oct      4.8
    #>  8 Australian Capital… Cafes, restaurants and cat… A3349849A   1982 Nov      5.4
    #>  9 Australian Capital… Cafes, restaurants and cat… A3349849A   1982 Dec      6.9
    #> 10 Australian Capital… Cafes, restaurants and cat… A3349849A   1983 Jan      3.8
    #> # … with 64,522 more rows

    # success!
    aus_retail %>% sample_n_keys(size = 10)
    #> # A tsibble: 4,266 x 5 [1M]
    #> # Key:       State, Industry [10]
    #>    State           Industry                        `Series ID`    Month Turnover
    #>    <chr>           <chr>                           <chr>          <mth>    <dbl>
    #>  1 Northern Terri… Cafes, restaurants and caterin… A3349377R   1988 Apr      1.7
    #>  2 Northern Terri… Cafes, restaurants and caterin… A3349377R   1988 May      2.6
    #>  3 Northern Terri… Cafes, restaurants and caterin… A3349377R   1988 Jun      2.8
    #>  4 Northern Terri… Cafes, restaurants and caterin… A3349377R   1988 Jul      3.1
    #>  5 Northern Terri… Cafes, restaurants and caterin… A3349377R   1988 Aug      3  
    #>  6 Northern Terri… Cafes, restaurants and caterin… A3349377R   1988 Sep      2.6
    #>  7 Northern Terri… Cafes, restaurants and caterin… A3349377R   1988 Oct      2.6
    #>  8 Northern Terri… Cafes, restaurants and caterin… A3349377R   1988 Nov      2.2
    #>  9 Northern Terri… Cafes, restaurants and caterin… A3349377R   1988 Dec      4.9
    #> 10 Northern Terri… Cafes, restaurants and caterin… A3349377R   1989 Jan      4  
    #> # … with 4,256 more rows

    # success!
    aus_retail %>% sample_frac_keys(size = 0.5)
    #> # A tsibble: 32,230 x 5 [1M]
    #> # Key:       State, Industry [76]
    #>    State           Industry        `Series ID`    Month Turnover
    #>    <chr>           <chr>           <chr>          <mth>    <dbl>
    #>  1 New South Wales Other retailing A3349872X   1982 Apr     178 
    #>  2 New South Wales Other retailing A3349872X   1982 May     182.
    #>  3 New South Wales Other retailing A3349872X   1982 Jun     174.
    #>  4 New South Wales Other retailing A3349872X   1982 Jul     180.
    #>  5 New South Wales Other retailing A3349872X   1982 Aug     178.
    #>  6 New South Wales Other retailing A3349872X   1982 Sep     176.
    #>  7 New South Wales Other retailing A3349872X   1982 Oct     180.
    #>  8 New South Wales Other retailing A3349872X   1982 Nov     193.
    #>  9 New South Wales Other retailing A3349872X   1982 Dec     260.
    #> 10 New South Wales Other retailing A3349872X   1983 Jan     182.
    #> # … with 32,220 more rows

<sup>Created on 2020-08-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by deanmarchiori** [comment]
Awesome, thanks for the package! Its part of my regular time series toolbox :)
